### DISCUSSION TITLE
Flatten does not work as expected when using a top-level field from a vector message

### DISCUSSION THREAD

**Discussion by teeohhem** [vectordotdev/vrl#387] [Q&A]
When using the flatten operation, referencing a field (that contains an object) from an event results in a no-op. This is because when you reference a field from a vector event, you are just given the value, which in this case, is just an object without a key. Therefore, obviously, there is nothing to flatten.

I personally would expect the flatten operation to "just work" and flatten with the key into the main event.

Here's an example: 

Event
```
{
    "timestamp": 1245542,
    "errors": {
        "code": 1234,
        "message": "invalid type",
        "class": "EventHelper"
    },
    "otherKey:" {
        "id": 1
    }
}
```

VRL
```
    flattened,err = flatten(.errors)
```

Expected: 
```
{
    "timestamp": 1245542,
    "errors.code": 1234,
    "errors.message": "Invalid type",
    "errors.class": "EventHelper",
    "otherKey:" {
        "id": 1
    }
}
```

Actual: 
No change to the event.

**Comment by pront** [comment]
Hi @teeohhem, I used your inputs [here](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLiA9IGZsYXR0ZW4oLikiLCJldmVudCI6eyJ0aW1lc3RhbXAiOjEyNDU1NDIsImVycm9ycyI6eyJjb2RlIjoxMjM0LCJtZXNzYWdlIjoiaW52YWxpZCB0eXBlIiwiY2xhc3MiOiJFdmVudEhlbHBlciJ9fSwiaXNfanNvbmwiOmZhbHNlLCJlcnJvciI6bnVsbH0%3D) and the result looks OK. 

Btw, your JSON event is missing some `,`s.

**Comment by teeohhem** [comment]
Hi @pront . Thanks for the response. I fixed my JSON event above. Yes, with `.` it works. But what if I just want to flatten `errors`? 

Here's the example: [here](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiIyBGb3IgaWxsdXN0cmF0aXZlIHB1cnBvc2VzLCBJIGFtIHJlcGxhY2luZyB0aGUgZXZlbnQgd2l0aCB0aGUgcmVzdWx0IGZyb20gZmxhdHRlblxuIyBJIHdvdWxkIGV4cGVjdCB0byBoYXZlIGFuIG9iamVjdCB3aXRoIGtleXMgZXJyb3JzLmNvZGUsIGVycm9ycy5tZXNzYWdlLCBlcnJvcnMuY2xhc3Ncbi4sZXJyID0gZmxhdHRlbiguZXJyb3JzKSIsImV2ZW50Ijp7InRpbWVzdGFtcCI6MTI0NTU0MiwiZXJyb3JzIjp7ImNvZGUiOjEyMzQsIm1lc3NhZ2UiOiJpbnZhbGlkIHR5cGUiLCJjbGFzcyI6IkV2ZW50SGVscGVyIn19LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D)

Here is my workaround/desired behavior: [here](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiZmxhdHRlbmVkX2ZpZWxkID0gZmxhdHRlbih7XCJlcnJvcnNcIjogLmVycm9yc30pXG4uID0gbWVyZ2UoLiwgZmxhdHRlbmVkX2ZpZWxkKVxuZGVsKC5lcnJvcnMpIiwiZXZlbnQiOnsidGltZXN0YW1wIjoxMjQ1NTQyLCJlcnJvcnMiOnsiY29kZSI6MTIzNCwibWVzc2FnZSI6ImludmFsaWQgdHlwZSIsImNsYXNzIjoiRXZlbnRIZWxwZXIifSwib3RoZXJLZXkiOnsiaWQiOjF9fSwiaXNfanNvbmwiOmZhbHNlLCJlcnJvciI6bnVsbH0%3D)

**Comment by pront** [comment]
For flattening the "errors" only see [this example](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLmVycm9ycyA9IGZsYXR0ZW4hKC5lcnJvcnMpXG4iLCJldmVudCI6eyJ0aW1lc3RhbXAiOjEyNDU1NDIsImVycm9ycyI6eyJjb2RlIjoxMjM0LCJtZXNzYWdlIjoiaW52YWxpZCB0eXBlIiwiY2xhc3MiOiJFdmVudEhlbHBlciJ9LCJvdGhlcktleSI6eyJpZCI6MX19LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D).

[This example](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLm9yaWdpbmFsID0gLlxuLnJlc3VsdDEgPSBmbGF0dGVuISguZXJyb3JzKVxuLnJlc3VsdDIgPSBmbGF0dGVuKC4pXG5cbiMgZGVsZXRpbmcgZm9yIHJlYWRhYmlsaXR5XG5kZWwoLnRpbWVzdGFtcClcbmRlbCguZXJyb3JzKSIsImV2ZW50Ijp7InRpbWVzdGFtcCI6MTI0NTU0MiwiZXJyb3JzIjp7ImNvZGUiOjEyMzQsIm1lc3NhZ2UiOiJpbnZhbGlkIHR5cGUiLCJjbGFzcyI6IkV2ZW50SGVscGVyIn19LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D) might help understand the differences.

**Comment by teeohhem** [comment]
@pront 
> For flattening the "errors" only see [this example](https://playground.vrl.dev/?state=eyJwcm9ncmFtIjoiLmVycm9ycyA9IGZsYXR0ZW4hKC5lcnJvcnMpXG4iLCJldmVudCI6eyJ0aW1lc3RhbXAiOjEyNDU1NDIsImVycm9ycyI6eyJjb2RlIjoxMjM0LCJtZXNzYWdlIjoiaW52YWxpZCB0eXBlIiwiY2xhc3MiOiJFdmVudEhlbHBlciJ9LCJvdGhlcktleSI6eyJpZCI6MX19LCJpc19qc29ubCI6ZmFsc2UsImVycm9yIjpudWxsfQ%3D%3D).

I'm arguing that if I want to flatten `.errors`, it would flatten the entire `.errors` object with each key having an `errors` key prefix.

In other words
flattening
```
{
    "errors": {
        "code": 1234,
        "message": "This is my message"
    }
}
```

should result in:
```
{
    "errors.code": 1234,
    "errors.message": "This is my message"
}
```
